bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,259 @@
+    protected abstract T_ID getIndentifier(T_OBJECT object);
+
+    /**
+     * Create a translated/localized copy of the given object.
+     *
+     * @param bundle the bundle where the config description with its translations originates from
+     * @param object the object to translate
+     * @param locale the target locale
+     * @return a translated copy of the given object or <code>null</code> if translation was not possible.
+     */
+    protected abstract T_OBJECT localize(Bundle bundle, T_OBJECT object, Locale locale);
+
+    /**
+     * Adds an object to the internal list associated with the specified module.
+     * <p>
+     * This method returns silently, if any of the parameters is {@code null}.
+     *
+     * @param bundle the module to which the object is to be added
+     * @param object the object to be added
+     */
+    public final synchronized void add(Bundle bundle, T_OBJECT object) {
+        if (object == null) {
+            return;
+        }
+        List<T_OBJECT> objects = acquireObjects(bundle);
+        if (objects == null) {
+            return;
+        }
+        objects.add(object);
+        // just make sure no old entry remains in the cache
+        removeCachedEntries(object);
+    }
+
+    /**
+     * Adds a {@link Collection} of objects to the internal list associated with the specified module.
+     * <p>
+     * This method returns silently, if any of the parameters is {@code null}.
+     *
+     * @param bundle the module to which the object is to be added
+     * @param objectList the objects to be added
+     */
+    public final synchronized void addAll(Bundle bundle, Collection<T_OBJECT> objectList) {
+        if (objectList == null || objectList.isEmpty()) {
+            return;
+        }
+        List<T_OBJECT> objects = acquireObjects(bundle);
+        if (objects == null) {
+            return;
+        }
+        for (T_OBJECT object : objectList) {
+            objects.add(object);
+            // just make sure no old entry remains in the cache
+            removeCachedEntries(object);
+        }
+    }
+
+    private List<T_OBJECT> acquireObjects(Bundle bundle) {
+        if (bundle == null) {
+            return null;
+        }
+        List<T_OBJECT> objects = bundleObjectMap.get(bundle);
+        if (objects == null) {
+            objects = new CopyOnWriteArrayList<T_OBJECT>();
+            bundleObjectMap.put(bundle, objects);
+        }
+        return objects;
+    }
+
+    /**
+     * Gets the object with the given key.
+     *
+     * @param key the key which identifies the object
+     * @param locale the locale
+     * @return the object if found, <code>null</code> otherwise
+     */
+    protected final T_OBJECT get(T_ID key, Locale locale) {
+        Collection<Entry<Bundle, List<T_OBJECT>>> objectList = bundleObjectMap.entrySet();
+        if (objectList == null) {
+            return null;
+        }
+        for (Entry<Bundle, List<T_OBJECT>> objects : objectList) {
```

Comment:
Why not using `... : bundleObjectsMap.entrySet()) {`?

entrySet shouldn't return null but an empty entry set.